storage/src/test/scala/coop/rchain/storage/regex/PathRegexUnitTests.scala

```diff
@@ -1,235 +1,180 @@
+              ("/test/route", "/test" :: Nil))
+
+    parse("/test/", noEndStrictOptions)
+      .struct("/test/" :: Nil)
+      .accept(("/test", Nil),
+              ("/test/", "/test/" :: Nil),
+              ("/test//", "/test/" :: Nil),
+              ("/test/route", "/test/" :: Nil))
   }
 
   "Combine modes" should "accept file path" in {
-    check(
-      "/test.json",
-      "/test.json" :: Nil,
-      ("/test.json", "/test.json" :: Nil)
-        :: ("/test.json.hbs", Nil)
-        :: ("/test.json/route", "/test.json" :: Nil)
-        :: Nil,
-      noEndStrictOptions
-    )
+    parse("/test.json", noEndStrictOptions)
+      .struct("/test.json" :: Nil)
+      .accept(("/test.json", "/test.json" :: Nil),
+              ("/test.json.hbs", Nil),
+              ("/test.json/route", "/test.json" :: Nil))
   }
 
   "Combine modes" should "work with named argument" in {
-    check(
-      "/:test",
-      PathToken(Some("test"), 0, Some('/'), Some('/'), false, false, false, """[^\/]+?""") :: Nil,
-      ("/route", "/route" :: "route" :: Nil)
-        :: ("/route/", "/route" :: "route" :: Nil)
-        :: Nil,
-      noEndStrictOptions
-    )
-    check(
-      "/:test/",
-      PathToken(Some("test"), 0, Some('/'), Some('/'), false, false, false, """[^\/]+?""") :: PathToken(
-        "/") :: Nil,
-      ("/route", Nil)
-        :: ("/route/", "/route/" :: "route" :: Nil)
-        :: Nil,
-      noEndStrictOptions
-    )
+    parse("/:test", noEndStrictOptions)
+      .struct(
+        PathToken(Some("test"), 0, Some('/'), Some('/'), false, false, false, """[^\/]+?""") :: Nil)
+      .accept(("/route", "/route" :: "route" :: Nil), ("/route/", "/route" :: "route" :: Nil))
+
+    parse("/:test/", noEndStrictOptions)
+      .struct(
+        PathToken(Some("test"), 0, Some('/'), Some('/'), false, false, false, """[^\/]+?""") :: pt(
+          "/") :: Nil)
+      .accept(("/route", Nil), ("/route/", "/route/" :: "route" :: Nil))
   }
 
   "Optional named parameter" should "work" in {
-    check(
-      "/:test?",
-      PathToken(Some("test"),
-                0,
-                Some('/'),
-                Some('/'),
-                optional = true,
-                repeat = false,
-                partial = false,
-                """[^\/]+?""") :: Nil,
-      ("/route", "/route" :: "route" :: Nil)
-        :: ("/route/nested", Nil)
-        :: ("/", "/" :: null :: Nil)
-        :: ("//", Nil)
-        :: Nil,
-      PathRegexOptions.default
-    )
+    parse("/:test?", PathRegexOptions.default)
+      .struct(
+        PathToken(Some("test"),
+                  0,
+                  Some('/'),
+                  Some('/'),
+                  optional = true,
+                  repeat = false,
+                  partial = false,
+                  """[^\/]+?""") :: Nil)
+      .accept(("/route", "/route" :: "route" :: Nil),
+              ("/route/nested", Nil),
+              ("/", "/" :: null :: Nil),
```

Comment:
I forgot to ask before: can we avoid using `null` here?

I ask because I try quite hard to avoid using `null` unless I am wrapping some underlying Java code where I have no choice, and usually then it will be converted to an `None` if I need to expose it from Scala.

Reply:
Well, currently PathRegex uses plain Java Regex, which returns null. I even made a simple test like: val rx = "abc(\\d+)?".r("num")
    "abc" match {
      case rx(num) => assert(num == null)
      case _ =>
    }
to ensure that scala regex macros is happy with nulls in similar cases.
When our Regex implementation will support capturing groups we likely implement them with Map("groupName" -> List[Capture]), or something similar (definitely not an Option[String]). Until that moment - naive replace args with Option(String) to avoid nulls will be bloody hard Pyrrhic victory on only 6 nulls - we will lose readability wrapping numerous arguments with Some(), and gain almost nothing.

Reply:
> Until that moment - naive replace args with Option(String) to avoid nulls will be bloody hard Pyrrhic victory on only 6 nulls - we will lose readability wrapping numerous arguments with Some(), and gain almost nothing.

😆 

Glad to hear you have considered this.  The `Map` solution sounds sensible to me.